From abef76bd780b3b379bf0ca54190a653be9efb166 Mon Sep 17 00:00:00 2001
From: Andrey Albershteyn <andrey.albershteyn@gmail.com>
Date: Tue, 26 Sep 2023 22:27:42 +0200
Subject: [PATCH] common: link .out file to the output directory

On the systems which source packages are read-only (e.g. NixOS) ln
will fail to link necessary .out file to the tests/ directory.
Therefore, link it to the final output directory, which has write
permission in any way, as .out.good.
---
 check         | 12 ++++++++----
 common/rc     |  2 +-
 tests/xfs/071 |  4 ++--
 3 files changed, 11 insertions(+), 7 deletions(-)

diff --git a/check b/check
index c897afbb41..58a441cea2 100755
--- a/check
+++ b/check
@@ -1042,7 +1042,11 @@
 		$timestamp && _timestamp
 		stop=`_wallclock`
 
-		if [ ! -f $seq.out ]; then
+		local out=$seq.out
+		if [ -f $seqres.out.good ]; then
+			out=$seqres.out.good
+		fi
+		if [ ! -f $out ]; then
 			_dump_err "no qualified output"
 			tc_status="fail"
 			_stash_test_status "$seqnum" "$tc_status"
@@ -1053,7 +1057,7 @@
 		# from `foo' to 'foo'. Filter old versions to match the new
 		# version.
 		sed -i "s/\`/\'/g" $tmp.out
-		if diff $seq.out $tmp.out >/dev/null 2>&1 ; then
+		if diff $out $tmp.out >/dev/null 2>&1 ; then
 			if [ "$tc_status" != "fail" ]; then
 				echo "$seqnum `expr $stop - $start`" >>$tmp.time
 				echo -n " `expr $stop - $start`s"
@@ -1062,13 +1066,13 @@
 		else
 			_dump_err "- output mismatch (see $seqres.out.bad)"
 			mv $tmp.out $seqres.out.bad
-			$diff $seq.out $seqres.out.bad | {
+			$diff $out $seqres.out.bad | {
 			if test "$DIFF_LENGTH" -le 0; then
 				cat
 			else
 				head -n "$DIFF_LENGTH"
 				echo "..."
-				echo "(Run '$diff $here/$seq.out $seqres.out.bad'" \
+				echo "(Run '$diff $here/$out $seqres.out.bad'" \
 					" to see the entire diff)"
 			fi; } | sed -e 's/^\(.\)/    \1/'
 			tc_status="fail"
diff --git a/common/rc b/common/rc
index cb91c00d24..cd942fb56e 100644
--- a/common/rc
+++ b/common/rc
@@ -3824,7 +3824,7 @@
 		print $result
 		' <$seqfull.cfg)
 	rm -f $1 || _fail "_link_out_file_named: failed to remove existing output file"
-	ln -fs $(basename $1).$suffix $1 || _fail "$(basename $1).$suffix: could not setup output file"
+	ln -fs $(realpath $1.$suffix) $seqres.out.good || _fail "$(basename $1).$suffix: could not setup output file"
 	test -r $1 || _fail "$(basename $1).$suffix: output file for feature set \"$2\" not found"
 }
 
diff --git a/tests/xfs/071 b/tests/xfs/071
index c4722f6718..930f82c364 100755
--- a/tests/xfs/071
+++ b/tests/xfs/071
@@ -24,9 +24,9 @@
 # link correct .out file
 rm -f $seqfull.out
 if [ "$bitsperlong" -eq 32 ]; then
-	ln -s $seq.out.32 $seqfull.out
+	ln -fs $(realpath $here/tests/xfs/$seq.out.32) $seqres.out.good
 else
-	ln -s $seq.out.64 $seqfull.out
+	ln -fs $(realpath $here/tests/xfs/$seq.out.64) $seqres.out.good
 fi
 
 _filter_io()
